syntax = "proto3";

import "google/protobuf/any.proto";

package tinc_protobuf;

enum MessageType {
  REQUEST = 0;
  REMOVE = 1;
  REGISTER = 2;
  CONFIGURE = 3;
  COMMAND = 4; // Commands are synchronous and one to one, so both ends block (or timeout) until COMMAND_REPLY is sent
  COMMAND_REPLY = 5;
  STATUS = 6;
  TINC_WORKING_PATH = 7;
  TINC_CLIENT_METADATA = 8;

  PING = 98;
  PONG = 99;

  BARRIER_REQUEST = 100;
  BARRIER_ACK_LOCK = 101;
  BARRIER_UNLOCK = 102;

  CLIENT_LOCK = 120;
  CLIENT_UNLOCK = 121;
  CLIENT_ACK_LOCK = 130;
  CLIENT_ACK_UNLOCK = 131;

  GOODBYE = 200;
}

enum ObjectType {
  PARAMETER = 0;
  PROCESSOR = 1;
  DISK_BUFFER = 2;
  DATA_POOL = 3;
  PARAMETER_SPACE = 4;
  GLOBAL = 5;
}

enum ProcessorType {
  DATASCRIPT = 0;
  CHAIN = 1;
  CPP = 2;
}

// This should be kept in sync with the C++ enum in ParamterSpaceDimension
enum ParameterType {
    PARAMETER_FLOAT = 0;
    PARAMETER_BOOL = 1;
    PARAMETER_STRING = 2;
    PARAMETER_INT32 = 3;
    PARAMETER_VEC3F = 4;
    PARAMETER_VEC4F = 5;
    PARAMETER_COLORF = 6;
    PARAMETER_POSED = 7;
    PARAMETER_MENU = 8;
    PARAMETER_CHOICE = 9;
    PARAMETER_TRIGGER = 10;

    PARAMETER_INT64 = 11;
    PARAMETER_INT16 = 12;
    PARAMETER_INT8 = 13;
    PARAMETER_UINT64 = 14;
    PARAMETER_UINT32 = 15;
    PARAMETER_UINT16 = 16;
    PARAMETER_UINT8 = 17;
    PARAMETER_DOUBLE = 18;
}

enum DataPoolTypes {
    DATAPOOLTYPE_JSON = 0;
    DATAPOOLTYPE_NETCDF = 1;
    DATAPOOLTYPE_USER = 2;
}

enum DiskBufferType {
    BINARY = 0;
    TEXT = 1;
    NETCDF = 2;
    JSON = 3;
    IMAGE = 4;
}

enum StatusTypes {
    UNKNOWN = 0;
    AVAILABLE = 1;
    BUSY = 2;
}


/// ################ TincMessage #####################
message TincMessage {
    MessageType messageType = 1;
    ObjectType objectType = 2;
    google.protobuf.Any details = 3;
}

message ObjectId {
  string id = 1;
}

message ClientMetaData {
  string clientHost = 1;
}

message DistributedPath {
  string filename = 1;
  string relativePath = 2;
  string rootPath = 3;

  string protocolId = 4;
}

// Parameter data message
message ParameterValue {
// This should be oneof, but 'repeated'is not allowed in oneof
    bool valueBool = 1;
    float valueFloat = 2;
    double valueDouble = 3;
    int32 valueInt32 = 4;
    uint32 valueUint32 = 5;
    int64 valueInt64 = 6;
    fixed64 valueUint64 = 7; // often used as bitfields
    string valueString = 20;
    repeated ParameterValue valueList = 30;
    uint32 nctype = 100;
}

// Registration Messages --------------------
message RegisterProcessor {
  string id = 1;
  ProcessorType type = 2;
  string inputDirectory = 3;
  repeated string inputFiles = 4;
  string outputDirectory = 5;
  repeated string outputFiles = 6;
  string runningDirectory = 7;
  string documentation = 50;
}

message RegisterDataPool {
  string id = 1;
  string parameterSpaceId = 2;
  string cacheDirectory = 3;
  DataPoolTypes type = 4;
  string documentation = 50;
}

message RegisterDiskBuffer {
  string id = 1;
  DiskBufferType type = 2;
  DistributedPath path = 3;
  string documentation = 50;
}

message RegisterParameterSpace {
  string id = 1;
  string documentation = 50;
}

message RegisterParameter {
  string id = 1;
  string group = 2;
  ParameterType dataType = 3;
  ParameterValue defaultValue = 4;
  string documentation = 50;
 }

// Configuration Messages
message ParameterSpaceValues {
  repeated string ids = 1;
  repeated ParameterValue values = 2;
}

enum ParameterConfigureType {
    VALUE = 0;
    MIN = 1;
    MAX = 2;
    SPACE = 3;
    SPACE_TYPE = 4;
}

// These values must conform to the values in the C++ and python libs
//enum ParameterSpaceConfigureType {
//    FLOAT_VALUE = 0;
//    ID = 1;
//    INDEX = 2;
//}

message ConfigureParameter {
  string id = 1;
  ParameterConfigureType configurationKey = 2;
  google.protobuf.Any configurationValue = 3;
}

enum ParameterSpaceConfigureType {
    ADD_PARAMETER = 0;
    REMOVE_PARAMETER = 1;
    CURRENT_TEMPLATE = 2;
    ROOT_PATH = 3;
    CACHE_PATH = 4;
    PS_DOCUMENTATION = 50;
}

message ConfigureParameterSpace {
  string id = 1;
  ParameterSpaceConfigureType configurationKey = 2;
  google.protobuf.Any configurationValue = 3;
}

message ConfigureProcessor {
  string id = 1;
  string configurationKey = 2;
  google.protobuf.Any configurationValue = 3;
}

enum DiskBufferConfigureType {
    CURRENT_FILE = 0;
    DB_DOCUMENTATION = 50;
}

message ConfigureDiskBuffer {
  string id = 1;
  DiskBufferConfigureType configurationKey = 2;
  google.protobuf.Any configurationValue = 3;
}

enum DataPoolConfigureType {
    SLICE_CACHE_DIR = 0;
    DATA_FILES = 1;
    DP_DOCUMENTATION = 50;
}
message DataPoolDataFiles {
  repeated string filename = 1;
  repeated string dimension = 2;
}

message ConfigureDataPool {
  string id = 1;
  DataPoolConfigureType configurationKey = 2;
  google.protobuf.Any configurationValue = 3;
}

// Remove Messages

message RemoveParameter {
  string id = 1;
  string group = 2;
}

// Commands

message Command {
    uint64 message_id = 1;
    ObjectId id = 2;
    google.protobuf.Any details = 3;
}

message CommandGenericPayload {
  bytes data = 1;
}

message CommandGenericAck {
}

message CommandErrorPayload {
  string error = 1;
}

// Request ParameterChoice elements
message ParameterRequestChoiceElements {
}

message ParameterRequestChoiceElementsReply {
    repeated string elements = 1;
}

// Request ParameterSpace current path
message ParameterSpaceRequestCurrentPath {
}

message ParameterSpaceRequestCurrentPathReply {
    string path = 1;
}

// Request ParameterSpace root path
message ParameterSpaceRequestRootPath {
}

message ParameterSpaceRequestRootPathReply {
    string path = 1;
}

// Request DiskBuffer current file slice
//message DiskBufferCommandFile {
//}

//message DiskBufferCommandFileReply {
//    string filename = 1;
//}


// Request DataPool slice
message DataPoolCommandSlice {
    string field = 1;
    repeated string dimension = 2;
}

message DataPoolCommandSliceReply {
    string filename = 1;
}

// Request DataPool current files
message DataPoolCommandCurrentFiles {
}

message DataPoolCommandCurrentFilesReply {
    repeated string filenames = 1;
}

// Status messages
message StatusMessage {
    ObjectId id = 1;
    StatusTypes status = 2;
}

message TincPath {
    string path = 1;
}
